cpp/ql/src/semmle/code/cpp/models/implementations/Strings.qll

```diff
@@ -0,0 +1,70 @@
+import semmle.code.cpp.models.interfaces.DataFlow
+import semmle.code.cpp.models.interfaces.Taint
+
+/**
+ * The `std::basic_string` constructor(s).
+ */
+class StringConstructor extends DataFlowFunction {
+  StringConstructor() {
+    this.hasQualifiedName("std", "basic_string", "basic_string")
+  }
+
+  override predicate hasDataFlow(FunctionInput input, FunctionOutput output) {
+    // flow from any constructor argument to return value
+    input.isInParameter(_) and
+    output.isOutReturnValue()
+  }
+}
+
+/**
+ * The standard function `std::string.c_str`.
+ */
+class StringCStr extends DataFlowFunction {
+  StringCStr() {
+    this.hasQualifiedName("std", "basic_string", "c_str")
+  }
+
+  override predicate hasDataFlow(FunctionInput input, FunctionOutput output) {
+    // flow from string itself (qualifier) to return value
+    input.isInQualifier() and
+    output.isOutReturnValue()
+  }
+}
+
+/**
+ * The standard function `operator<<`, for example on `std::stringstream`.
+ */
+class InsertionOperator extends TaintFunction {
+  InsertionOperator() {
+    this.hasQualifiedName("std", "operator<<")
+  }
+
+  override predicate hasTaintFlow(FunctionInput input, FunctionOutput output) {
+    (
+      // flow from second argument to return value
+      input.isInParameterPointer(1) and
+      output.isOutReturnValue()
+    ) or (
+      // flow from return value to first argument
+      // (this is a bit odd; another way to think of it is the sink effectively
+      //  flowing from the first argument to the return value)
+      input.isInReturnValue() and
+      output.isOutParameterPointer(0)
```

Comment:
In the case of:
```
ss3 << "123" << source();
```
i.e.
```
(ss3 << "123") << source();
```
taint needs to flow from the bracketed expression to `ss3` somehow (in actuality I think the thing that's flowing is a reference to the sink, in the other direction, which is why my attempt at solving this looks so backwards).

Reply:
I think I understand the problem now.

The `ss3` argument should have a post-update node (a `DefinitionByReferenceNode`) since it's passed by mutable reference into `operator<<`. Does it work to add flow from `(ss3 << "123")` to the post-update node of `ss3`?

Reply:
To be more precise, I think these are the steps we want in the model of `operator<<`:
- Data flow: `isInParameter(0) --> isOutReturnValue()`
- Taint: `isInParameterPointer(1) --> isOutReturnPointer()`

We should really write some documentation in `FunctionInputsAndOutputs.qll`. Maybe the documentation is missing because we're still figuring out what each of the input/output kinds ought to mean for taint and data flow. I _think_ that for pointer/reference types, the `Value` kinds (like `isOutReturnValue`) refer refer to the pointer itself, and the `Pointer` kinds (like `isOutReturnPointer`) refer to the contents that are pointed to. If that's the case, we should rename `Pointer` to `Content` or `Pointee` or something.

When a model specifies "parameter value" to "return value" _data flow_, I think the data flow library should add both forward flow between the pre-update nodes and backwards flow between the post-update nodes. I had at one point a plan to generalise https://github.com/Semmle/ql/pull/1766/files#diff-255652465f915d69e8f3adeed3e96f11R238 to add such backwards post-update flow everywhere that the right post-update nodes existed, but we decided that it was more sensible to add these cases one at a time. As an example of why we want such flow, consider an identity function on reference types: `int& id(int &x) { return x; }`. Adding backwards post-update flow would let us model `id(id(a)) = source;` as an assignment to `a` (assuming we have flow from pre-update RHS to post-update LHS of `=`).

When a model specifies flow to a `Pointer` output (the contents pointed to), I think the data flow (or taint flow) library should interpret that as flow into the appropriate _post-update_ node. In this case, flow to `isOutReturnPointer()` goes to the post-update node for the `operator<<` call. Such a post-update node should be made to exist when the call itself is being used as an argument to a function that takes a pointer argument. It doesn't exist today -- that's one of the coverage gaps we talked about in #1900.

With this mapping in place between the models and the data flow library, I think it's going to work. Then we get the following data-flow edges for your example with `(ss3 <<₁ "123") <<₂ source();`, where `==>` is taint, and `-->` is data flow:

```
# Edges for <<₁:
pre-ss3 --> pre-<<₁
post-<<₁ --> post-ss3 # backwards post-update flow
pre-"123" ==> post-<<₁

# Edges for <<₂:
pre-<<₁ --> pre-<<₂
post-<<₂ --> post-<<₁ # backwards post-update flow
pre-source() ==> post-<<₂
```

In that graph, there is a path from `pre-source()` to `post-ss3` as desired.

Reply:
I've discussed this with @aschackmull. The Java library has a `StringBuilder` class that is more or less equivalent to `std::stringstream`. Calls like `sb.append(s1).append(s2)` are modelled by introducing a concept of the "initial qualifier" on a chain of calls -- `sb` in this case. For C++ we'd prefer to do this modelling only on the `Function` and not on the `Call`, and I think that's possible.

> When a model specifies flow to a `Pointer` output (the contents pointed to), I think the data flow (or taint flow) library should interpret that as flow into the appropriate _post-update_ node.

The post-update node I mentioned above may not exist. I'd like to try and investigate whether a `Call` can be its own post-update node.

Reply:
I think the two-rule modelling of `operator<<` that I proposed in https://github.com/Semmle/ql/pull/1637#discussion_r324199025 was too "clever", relying on flow through post-order nodes of calls that happened to exist in some cases but only by coincidence. Here's how I now think we should model `operator<<`, using the terminology of #1938. It's the same as what I wrote in https://github.com/Semmle/ql/pull/1637#discussion_r313518288, but this time I'm being precise about what's pre and what's post.
- Data flow: `isParameter(0) --> isReturnValue()`. On `a << b`, this should give data flow from `pre-a` to `pre-<<` and from `post-<<` to `post-a`.
- Taint: `isParameterDeref(1) --> isReturnValueDeref()`. On `a << b`, this should give taint flow from `pre-b` to `pre-<<`. I'd previously claimed this should go to `post-<<`, but that doesn't exist in general for a call, and the "post-result" of a call is just the result.
- Taint: `isParameterDeref(1) --> isParameterDeref(0)`. On `a << b`, this should give taint flow from `pre-b` to `post-a`.

That gives us the following edges for `(ss3 <<₁ "123") <<₂ source();`, where there exists a path from `pre-source()` to `post-ss3`.

```
# Edges for <<₁:
pre-ss3 --> pre-<<₁ # due to isInParameter(0) --> isOutReturnValue()
post-<<₁ --> post-ss3 # backwards post-update flow of the above
pre-"123" ==> pre-<<₁ # due to isParameterDeref(1) --> isReturnValueDeref()
pre-"123" ==> post-ss3 # due to isParameterDeref(1) --> isParameterDeref(0)

# Edges for <<₂:
pre-<<₁ --> pre-<<₂ # due to isInParameter(0) --> isOutReturnValue()
# post-<<₂ --> post-<<₁ # (there is no post-<<₂ node)
pre-source() ==> pre-<<₂ # due to isParameterDeref(1) --> isReturnValueDeref()
pre-source() ==> post-<<₁ # due to isParameterDeref(1) --> isParameterDeref(0)
```